---
title: "Vignette Preprocessing HT12-v4 Gene Expression Data"
author: "Holger Kirsten & Carl Beuchel"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
---
 
 
## Initialisation for a clean start 

```{r}
knitr::opts_chunk$set(cache = FALSE, results = "markup", echo=T , fig.width = 9, fig.height = 6, eval = F)

```

  
```{r}
rm(list = ls())
# .libPaths("/mnt/ifs1_projekte/genstat/07_programme/rpackages/forostar/")


require(rmarkdown)
require(lumi)
require(ggplot2)
require(data.table)
require(knitr)
require(pander)
pander::panderOptions('table.split.table', Inf)


 require(HT12ProcessoR)




options(stringsAsFactors=FALSE)
```


This example shows a typical preprocessing of ILLUMINA HT12v4 data using the package HT12ProcessoR. This can be applied for a few samples up to thousands of samples. Details on the Illumina technology can be found here: http://support.illumina.com/content/dam/illumina-marketing/documents/products/datasheets/datasheet_gene_exp_analysis.pdf 
  
## Before Starting: Organize input files and customize project-specific parameters
  
  
  * Start with Data from Illuminas GenomeStudio. Data should be non-normalized, non-background-substracted. Illumina organises expression data in filesets consisting of three files, a sample file with gene probe data, a control files with control probe data and a sample file with individidual-related information
  
  * A tab-delimeted file should be created specifying location of these three files. Following columns are required:
  
      + `fileset_id`: a short ID for a fileset
      + `con_f`: full path and filename of the control probe data file
      + `nobkgd_f`: full path and filename of the gene probe data file
      + `sample_f`: full path and filename of the control probe data file
      
  * This packages comes with an included example (a subset with 96 Chips from GEO accession: GSE65907). This data can be found in `r paste0(path.package("HT12ProcessoR"), "/", "inst")`
  
  * The file should be saved in the folder intended for preprocessing the data

```{r}

datadirectory = path.package("HT12ProcessoR")
# datadirectory = '/mnt/ifs1_projekte/genstat/07_programme/rtools/HT12ProcessoR/inst/'

# input = data.table(fileset_id = c('f2226-2270', '334-2184'),
# con_f = c(paste0(datadirectory,"/ControlProbeProfile_2226-2270.txt"),paste0(datadirectory,"/ControlProbeProfile_334-2184.txt")),
# nobkgd_f = c(paste0(datadirectory,"/Einzelanalyse_nonorm_nobkgd_2226-2270.txt"),paste0(datadirectory,"/Einzelanalyse_nonorm_nobkgd_334-2184.txt")),
# sample_f = c(paste0(datadirectory,"/SamplesTable_2226-2270.txt"),paste0(datadirectory,"/SamplesTable_334-2184.txt")))

 input = data.table(fileset_id = c('f34-21', '334-2184'), 
                     con_f = c(paste0(datadirectory,"/Run12_ControlProbeProfile_590-651.txt"),paste0(datadirectory,"/ControlProbeProfile_HSS106-HSS155.txt")),
                     nobkgd_f = c(paste0(datadirectory,"/Run12_nonorm_nobkgd_590-651.txt"),paste0(datadirectory,"/Einzelanalyse_nonorm_nobkgd_HSS106-HSS155.txt")),
                     sample_f = c(paste0(datadirectory,"/Run12_SamplesTable_590-651.txt"),paste0(datadirectory,"/SamplesTable_HSS106-HSS155.txt")))


prepro_folder = "R:/genstat/07_programme/rtools/HT12ProcessoR/_archive/example005"
dir.create(prepro_folder, recursive = T)

input_filename = paste0(prepro_folder, "/s01a_file_overview.txt")

write.table(input,input_filename ,quote=F, col.names=T, row.names=F, sep="\t")

```

Next, customize project-specific variables in the parameter file. This file is used to provide most relevant parameters in a single place for the individual preprocessing functions. Parameters will also be explained later. An example is provided with this package

```{r}
myparams = read.delim(paste0(datadirectory, "/input_parameter_010.txt"), as.is = T)

```

Project-specific parameters that should be customized initially are

- `file_names_of_files_and_folders`: Location of the files to preprocess. This is the `input_filename` created in the previous step
- `datafolder_results`: Directory, where the results of preprocessing can be stored
```{r}

myparams[ myparams$variable == "file_names_of_files_and_folders", "value"] = input_filename
myparams[ myparams$variable == "datafolder_results", "value"] = paste0(prepro_folder, "/tosent/")

#in Case all samples have spiked-in ERCC controls
#myparams[ myparams$variable == "excludeERCC", "value"] = FALSE
#myparams[ myparams$variable == "controlparameters2check", "value"] = "hybrid_low, hybrid_med,  string_pm, string_mm, biotin, negative, ercc_1, ercc_2, ercc_3, ercc_4, ercc_5"

## if removeBatchEffects() does not work, this might be due to to strongly correlated controlparameters2check
#myparams[ myparams$variable == "controlparameters2check", "value"] = "hybrid_low, hybrid_med,  string_pm, string_mm, biotin, negative, ercc_1,  ercc_3, ercc_4, ercc_5"


myparamfile = paste0(prepro_folder, "/input_parameter_010.txt")
write.table(myparams,myparamfile ,quote=F, col.names=T, row.names=F, sep="\t")

```

  
## Step 1 - function `checkExtractChipsamples`: Check given expression files  and extract individuals

This step does the following:
  
  * checks files specified in myparamfile for:
  + are expected files present
  + does each file have information for each sample
  + multiple identical IDs in different expression files present, if yes, this will be resolved according to parameter 
  + properly formated sentrx IDs for all samples present
  + sentrix IDs for all samples present
  + 47323 ILMN gene expression probes present in all files
  + 883 ILMN control probes present in all files
  
  
  * create list-object  of class **HT12prepro** including a slot with sample-related attributes of the current processing-stage named `$chipsamples`

```{r }

prepro_ht12 = checkExtractChipsamples(paramfile =myparamfile)

pander(prepro_ht12$history)
class(prepro_ht12)

pander(ht(prepro_ht12$chipsamples,1))

```

## Step 2  Designate individuals and subgroups for preprocessing and redefine batches

* Samples can be excluded if they should not be included in preprocessing, e.g. if they are genotyped on the same chip but are from a totally different project. For this, column `old_ID` in prepro_ht12$chipsamples can be used, which reports the original ID used in the raw data from column `Sample ID` of the sample-file from ILLUMINA

* Major subgroups of the data should be specified, that should treated independet. A typical example is if data from different tissue is preprocessed together

* This will force the functions, that the number of detected and expressed probes as well as the Euclidian distance of all expressed probes (used to define atypical samples) are calculated subgroup-specific. Optionally, batch effect correction by function `removeBatchEffects()` can be done by preserving the subgroup contrast

* Additionally, up to three (nested) batches can be specified, that should reflect processing. 
    + Default First-level batch is `Sentrix.Barcode`, the Chip-ID. 
    + Typically, several chips are processed together, this variable is defined as `processingbatch`. The default is the column `fileset_id` from the above file `input_filename`. 
    + Finally, sometimes several runs behave differently, these might be specified in the variable `strangebatch`, default here is 0.


```{r}
ht(prepro_ht12$chipsamples,2)


## Exclude Samples

prepro_ht12$chipsamples = prepro_ht12$chipsamples[ ! prepro_ht12$chipsamples$old_ID  %in%  c("614", "638", "HSS 108", "HSS 112", "HSS 115", "HSS 125", "HSS 127", "HSS 133", "HSS 137", "HSS 138", "HSS 146", "HSS 151","HSS 155"),]
## Define Subgroups
prepro_ht12$chipsamples$subgroup = "PBMC"
table(prepro_ht12$chipsamples$subgroup)

xtabs(~prepro_ht12$chipsamples$Sentrix.Barcode + prepro_ht12$chipsamples$processingbatch)

xtabs(~prepro_ht12$chipsamples$strangebatch + prepro_ht12$chipsamples$processingbatch)


## Quick Check for major differences based on Attributes from the provided ILLUMINA-sample-files
initial_pca = calcInitialPCA(prepro_ht12)

p1 = ggplot(initial_pca$scores, aes(PC1, PC2, pch = subgroup , col = processingbatch , size = Detected.Genes..0.01., alpha = in_study, label = paste(old_ID, new_ID, sep = "\n"))) + geom_point() + scale_alpha_manual(values = c(0.7, 0.3))
p1

## interactive visualisation 
require(plotly)
p1 %>% ggplotly()

```


## Step 3 - function `createExpressionSet()`: Create an expression set and related information from data

* expression information that is scattered across many files is united in a single expression set object. This includes control probe information. 

* Within the updated HT12prepro-object among others, following slots are created:
    + a slot with  sample-related attributes of the current processing-stage named `$chipsamples`
    + a slot with the R-object holding expression data used to create the expression set named `$rawdataWOcons_joined`
    + a slot with the R-object holding control data used to create the expression set named `$rawdataOnlycons_joined`
    + a slot with the R-object holding control data used to create the expression set named `$rawdataOnlycons_joined`
    + a slot with expression-set expression data excluding control data named `$all_nobkgd_eset`
    + a slot with expression-set expression data including control data named `$total_nobkgd_eset`
    + a slot with the history of the commands named `$history`

* Integrity checks are done:
    + that gene expression data and control expression data for all samples imported
    + check for NAs present in gene expression and control expression data 
    + check for negative gene expression values in gene expression and control expression data, if found, replaced with the lowest expression value found
    + check for similar values of ERCC probes and housekeeping probes in gene-expression and control-expression data as Illumina reports them in both of the corresponding files
  
```{r}
prepro_ht12 = createExpressionSet(prepro_ht12, paramfile = myparamfile)

## Have a look:
kable(prepro_ht12$history)
prepro_ht12$all_nobkgd_eset
hh(prepro_ht12$rawdataWOcons_joined)
hh(detection(prepro_ht12$all_nobkgd_eset))

prepro_ht12$total_nobkgd_eset
hh(prepro_ht12$rawdataOnlycons_joined)
hh(detection(prepro_ht12$total_nobkgd_eset))

```

## step 4 - function `filterLowExpressed`: Filter badly expressed individuals & mark badly expressed genes

* filter samples for extreme number of expressed transcripts and annotate transcripts for extreme numbers of samples not found to have the transcript expressed
* filtering samples for extreme numbers of expressed transcripts uses parameter `filter1ind_expressedGenes`as threshold for the number of interquantile ranges of the number of detected probes to be tolerated. The definition of detection is related to a ILLUMIN-defined detection p-value of 0.01
* annotating transcripts for having extreme numbers of samples where ´expressed´ uses parameter `filter1probes_expressedProbes`as threshold, default is 0.05

 * Within the updated HT12prepro-object among others, 
    + the slot with  sample-related attributes of the current processing-stage named `$chipsamples` is updated 
    + a slot with the detailed probe-related expression level informationis created named `$genesdetail
    + the slot with the history of the commands named `$history`` is updated
    

```{r }
prepro_ht12 = filterLowExpressed(prepro_ht12, paramfile = myparamfile)
kable(prepro_ht12$history)

head(prepro_ht12$genesdetail)
setDT(prepro_ht12$chipsamples)
prepro_ht12$chipsamples[,.N,.(in_study, reason4exclusion)]

setDF(prepro_ht12$chipsamples)

```


## step 5 - functions `transformNormalizeHT12object()` and `filterTechnicallyFailed()` Transform & normalize data & filter samples for atypical control values

* all expression probes (incl. control probes) of the still valid samples are log2 transformed and normalized (default: method quantile, via parameter `normalisation_method` , method `rsn` is also available)
* meaningfull and not heavily correlated QC-parameters have to be identified to define a single combined measure representing technical performance. Default is  'hybrid_low, hybrid_med,  string_pm, string_mm, biotin, negative' (if spiked in erccc is used in ALL CHIPS use also 'ercc_1, ercc_2, ercc_3, ercc_4, ercc_5', if spiked in artificial polyadenylated RNAs from Bacillus subtilis is used, also 'labeling'). In case of population based studies within a single tissue also 'housekeeping, Detected.Genes..0.01.' can be included)

* Mahalanobis distance of this control parameters is calculated and outlyers are identified defined as having a greated distance 3 interquantile ranges. This number can be changed via parameter `filter2ind_atypischIlmnKontroll`

 * Within the updated HT12prepro-object among others, 
    + the slot with  sample-related attributes of the current processing-stage named `$chipsamples` is updated 
    + a slot with an expression set with normalized and transformed data excluding control probe information is created: `$all_nobkgd_eset_ql`  
    + a slot with an expression set with normalized and transformed data including control probe information is created: `$total_nobkgd_eset_ql`    + the slot with the history of the commands named `$history`` is updated

* Several QC plots are shown displaying ILLUMINAS QC-parameter. See also http://www.dkfz.de/gpcf/illumina_beadchips.html for details on Illumina control probes 

* Within this process, Illuminas QC parameters are recalculated using the transformed and normalized data

* creates file __obj/s06_file_sampleattributes_in_combined_total_eset_expr_mahal.txt__, containing the sample related attributes of the current proccessing stage

```{r }
prepro_ht12 = transformNormalizeHT12object(prepro_ht12, paramfile = myparamfile)

prepro_ht12 = filterTechnicallyFailed(prepro_ht12, paramfile = myparamfile)
kable(prepro_ht12$history)
```


## step 6 - functions `filter4MinBatchsize()`, `transformNormalizeHT12object()`, and `removeBatchEffects()`: Transform and normalize still valid samples and adjust for batch-effects

* all expression probes (incl. control probes) of all still valid samples are log2-transformed and normalized (method quantile), again, as this procedure depends on the set of included samples

* Within the updated HT12prepro-object among others, 
    + the slot with  sample-related attributes of the current processing-stage named `$chipsamples` is updated 
    + a slot with an expression set with normalized and transformed and batch-corrected data including control probe information is created: `$total_nobkgd_eset_ql_combat`    
    + the slot with the history of the commands named `$history`` is updated

* Several QC plots are shown, see also function ComBat() from the R/package sva for details

```{r }

prepro_ht12 = filter4MinBatchsize(prepro_ht12, paramfile = myparamfile)


prepro_ht12 = transformNormalizeHT12object(prepro_ht12, paramfile = myparamfile) ## only necessary if samples were removed in previous step

prepro_ht12 = removeBatchEffects(prepro_ht12, paramfile = myparamfile)
kable(prepro_ht12$history)

```

## step 7 - function `checkBatchEffects()`: Check for remaining batch effects 

* Identification of transcripts where association with batches is still stronger than expected by chance after batch-adjustment. Stronger than chance (i.e. overinflation) is defined as association stronger p-value after Bonferroni correction

* All Batches defined in step 2 will be tested, i.e. `Sentrix.Barcode`,  `processingbatch`, and - if defined - `strangebatch`

* Probes are not filtered but annotated when found to be overinflated


* Within the updated HT12prepro-object among others, 

    + the slot with  probe-related attributes of the current processing-stage named `$genesdetail` is updated 
    + the slot with the history of the commands named `$history`` is updated

* QQ plots of association are shown


```{r }
prepro_ht12 = checkBatchEffects(prepro_ht12, paramfile = myparamfile)
ht(prepro_ht12$genesdetail,2)
kable(prepro_ht12$history)
```


## step 8 - function `filterAtypicalExpressed()`: Identify samples with atypical expression values

* Filter samples for atypical gene expression levels. This is defined as atypical large Euclidian distance of all expressed, not batch-associated expression probes

* Outlyers are identified defined as having a greated distance than 3 interquantile ranges. This number can be changed via parameter `filter2ind_atypischEuklid` 

* Note that sometimes it might be meaningful to repeat correction for batch association after excluding atypical samples identified in this step

* Within the updated HT12prepro-object among others, 
    + the slot with  sample-related attributes of the current processing-stage named `$chipsamples` is updated 
    + the slot with the history of the commands named `$history`` is updated

```{r }
prepro_ht12 = filterAtypicalExpressed(prepro_ht12, paramfile = myparamfile)
ht(prepro_ht12$chipsamples,2)
kable(prepro_ht12$history)
```


## step 9 - function `visualizePreprocessing()` and `comparePCBeforeAfterPrePro()`: visualisation of Quality Control measures

* visualisation of control features before and after preprocessing in several plots
* raw data before preprocessing is log2 transformed for a more meaningful comparison
* PCA on expression data is also done using good samples surviving preprocessing. Following PCAs are calculated and shown as figures:
    + PCA of all transcripts before preprocessing, including additional scaling
    + PCA of gene expression after preprocessing using only expression probes that are expressed, not overinflated regarding batch affects and beeing specific to the human genome when considering a remapping approach according to Barbosa-Moralis et al. (2010)

* Within the updated HT12prepro-object among others, 
    + plots are also stored in slot `$`
    + the slot with the history of the commands named `$history`` is updated


```{r }

prepro_ht12 = visualizePreprocessing(prepro_ht12, paramfile = myparamfile)

prepro_ht12 = comparePCBeforeAfterPrePro(prepro_ht12, paramfile = myparamfile)
kable(prepro_ht12$history)
```

## step 10 - function `writeFilesTosent()`: Write preprocessed data

* directory of the resulting preprocessed data was specified in step *Before Starting*. Individual filenames can be customized by specifying parameters 

    + `file_sampleannot_final.txt`: default is	sampleannot_HT12v4.txt --> annotation of samples
    + `file_probeannot_final.txt`: default is		probeannot_HT12v4.txt --> annotation of probes
    + `file_annot_final.xlsx`: default is		sampleUNDprobeannot_HT12v4.xlsx --> annotation of samples and probes as excel file
    + `file_final_expression_set`: default is		expressionset_preprocessed.Rdata --> Preprocessed data as expression set
    + `file_final_expression_matrix`: default is	expressionmatrix_preprocessed.Rdata --> Preprocessed data as R-Matrix
    + `file_all_transcripts_good_incl_remapping_ok`: default is	all_transcripts_good_incl_remapping_ok.txt --> IDs of probes that are considered expressed in all subgroups, have no remaining batch effect stronger than expected by chance, and are ok given a remapping approach of Barbosa-Moralis et al. (2010). Not that probes with a remaining batch effect might still be included in association analysis if no nominal significant correllation is found between the batch and the analysis variable of interest.
    + `file_final_expression_matrix_allProbes.txt`: default is	expressionmatrix_preprocessed.txt --> Filename of expression matrix as tab delimited text file after finished preprocessing


* Parameter `file_renaming_samples_tosent` can specify a filename where a column named `oldname` and `newname` is provided in order to rename sample. Column oldname must match the names of the samples in column new_ID  in the `$chipsamples` data.frame included in the HT12 object. Can be empty if no renaming is required 

* note that expression values are not residualised for any covariates e.g. age, sex, or any other phenotypic variable or any principal component of expression data

```{r}
## Set parameter to use provided renaming file
renaming_fn = paste0(datadirectory, "/renamingsamples.txt")
myparams[ myparams$variable == "file_renaming_samples_tosent", "value"] = renaming_fn
write.table(myparams,myparamfile ,quote=F, col.names=T, row.names=F, sep="\t")

prepro_ht12 = writeFilesTosent(prepro_ht12, paramfile = myparamfile)
ht(prepro_ht12$chipsamples,1)
prepro_ht12$history

```

It might be helpful, to save the HT12-object to preserve detailed information on preprocessing

```{r}
save(prepro_ht12, file = paste0(prepro_folder,"/prepro_example1_HT12object.RData"))

```

## step 11 - function `createTextForMethods()`: Information for Material and Methods section of a manuscript

This function creates an extended and a short version of the preprocessing done for section Material and Methods of a publication
**Please slightly rephrase text as most journals require unique phrasing even in materials & methods section**



```{r results="asis"}
mm_text = createTextForMethods(prepro_ht12, paramfile = myparamfile)
cat('### Extended Version')
cat(mm_text$extended_version)
cat('### Short Version')
cat(mm_text$short_version)
```



**_Session Info_**

```{r sesionifno, results='markup'}
sessionInfo()
```


